src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,180 @@
+                if (metadata == null)
+                {
+                    // Can match any request
+                    continue;
+                }
+
+                string metadataHeaderName = metadata.HeaderName;
+                var metadataHeaderValues = metadata.HeaderValues;
+
+                bool matched = false;
+                if (httpContext.Request.Headers.TryGetValue(metadataHeaderName, out var requestHeaderValues))
+                {
+                    if (metadataHeaderValues.Count == 0)
+                    {
+                        // We were asked to match as long as the header exists, and it *does* exist
+                        matched = true;
+                    }
+                    else
+                    {
+                        var comparisonFunc = GetComparisonFunc(metadata.HeaderValueMatchMode);
+                        var stringComparison = metadata.HeaderValueStringComparison;
+                        for (int j = 0; j < metadataHeaderValues.Count; j++)
+                        {
+                            for (int k = 0; k < requestHeaderValues.Count && k < maxRequestHeadersToInspect; k++)
+                            {
+                                if (comparisonFunc(requestHeaderValues[k], metadataHeaderValues[j], stringComparison))
+                                {
+                                    matched = true;
+                                    break;
+                                }
+                            }
+
+                            if (matched)
+                            {
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                if (!matched)
+                {
+                    candidates.SetValidity(i, false);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static Func<string, string, StringComparison, bool> GetComparisonFunc(HeaderValueMatchMode matchMode)
+        {
+            return matchMode switch
+            {
+                HeaderValueMatchMode.Prefix => Prefix,
+                _ => Exact,
+            };
+
+            static bool Exact(string a, string b, StringComparison comparison) => string.Equals(a, b, comparison);
+            static bool Prefix(string a, string b, StringComparison comparison) => a != null && b != null && a.StartsWith(b, comparison);
+        }
+
+        private static bool AppliesToEndpointsCore(IReadOnlyList<Endpoint> endpoints)
+        {
+            return endpoints.Any(e =>
+            {
+                var metadata = e.Metadata.GetMetadata<IHeaderMetadata>();
+                return !string.IsNullOrEmpty(metadata?.HeaderName);
+            });
+        }
+
+        private class HeaderMetadataEndpointComparer : EndpointMetadataComparer<IHeaderMetadata>
+        {
+            protected override int CompareMetadata(IHeaderMetadata x, IHeaderMetadata y)
```

Comment:
I think there needs to be extra logic here about which header is most specific.

For example, an exact match is more specific than a prefix match.